File: backend/mcp-code-intelligence-kotlin/bin/main/com/codeintel/analyzers/entrypoints/detectors/CLIDetector.kt
/**
 * KSA-162: CLI Command Detector.
 */
package com.codeintel.analyzers.entrypoints.detectors

import com.codeintel.analyzers.entrypoints.*

private val CLI_PATTERNS = listOf(
    "commander" to ".command(",
    "yargs" to ".command(",
    "click" to "@click.command",
    "argparse" to "add_subparsers",
    "cobra" to "cobra.Command",
    "clap" to "#[command",
)

class CLIDetector {

    fun detect(symbols: List<SymbolInput>, source: String): List<EntryPoint> {
        val cliFramework = CLI_PATTERNS.find { source.contains(it.first) } ?: return emptyList()
        return symbols.filter { sym ->
            val decorators = sym.decorators ?: emptyList()
            decorators.any { "command" in it || "cli" in it } ||
                "command" in sym.name || "cmd" in sym.name || sym.name.startsWith("cli_")
        }.map { sym ->
            EntryPoint(
                symbolId = sym.id, symbolName = sym.name,
                filePath = sym.filePath, startLine = sym.startLine,
                entryType = EntryType.CLI_COMMAND, framework = cliFramework.first,
                eventName = sym.name, confidence = Confidence.Medium,
            )
        }
    }
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/analyzers/entry_points/detectors/cli_detector.py
"""KSA-162: CLI Command Detector."""
from __future__ import annotations
from .. import EntryPoint, EntryType, Confidence, SymbolInput

CLI_PATTERNS = [
    ("commander", ".command("), ("yargs", ".command("),
    ("click", "@click.command"), ("argparse", "add_subparsers"),
    ("cobra", "cobra.Command"), ("clap", "#[command"),
]


class CLIDetector:
    def detect(self, symbols: list[SymbolInput], source: str) -> list[EntryPoint]:
        framework = next((p for p in CLI_PATTERNS if p[0] in source), None)
        if not framework:
            return []
        results = []
        for sym in symbols:
            decs = sym.decorators or []
            is_cli = (any("command" in d or "cli" in d for d in decs) or
                      "command" in sym.name or "cmd" in sym.name or sym.name.startswith("cli_"))
            if is_cli:
                results.append(EntryPoint(
                    symbol_id=sym.id, symbol_name=sym.name,
                    file_path=sym.file_path, start_line=sym.start_line,
                    entry_type=EntryType.CLI_COMMAND, framework=framework[0],
                    event_name=sym.name, confidence=Confidence.Medium,
                ))
        return results
